### ISSUE TITLE
civirules preventing drush and cv from clearing cache

### ISSUE THREAD

**Issue by lsh-0** [CiviCooP/org.civicoop.civirules#180]
I'm trying to automate the installation and removal of custom extensions for our instance of civi. We use `cv` and `drush` to get this done.

We have a problem where the initial installation of civirules works fine, but then when calling `cv flush` or `drush cc civicrm` or attempting to install the extension again or run any db migrations for extensions, we get this error:

```php
$ cv flush
Flushing system caches
Error: API Call Failed: Array
(
    [entity] => System
    [action] => flush
    [params] => Array
        (
            [triggers] => 1
            [debug] => 1
            [version] => 3
        )

    [result] => Array
        (
            [trace] => #0 /srv/crm-project/crm/sites/all/modules/civicrm/CRM/Core/ManagedEntities.php(234): CRM_Core_ManagedEntities->onApiError('CiviRuleTrigger', 'create', Array, Array)
#1 /srv/crm-project/crm/sites/all/modules/civicrm/CRM/Core/ManagedEntities.php(179): CRM_Core_ManagedEntities->insertNewEntity(Array)
#2 /srv/crm-project/crm/sites/all/modules/civicrm/CRM/Core/ManagedEntities.php(141): CRM_Core_ManagedEntities->reconcileEnabledModule(Object(CRM_Core_Module), Array)
#3 /srv/crm-project/crm/sites/all/modules/civicrm/CRM/Core/ManagedEntities.php(122): CRM_Core_ManagedEntities->reconcileEnabledModules()
#4 /srv/crm-project/crm/sites/all/modules/civicrm/CRM/Core/Invoke.php(395): CRM_Core_ManagedEntities->reconcile()
#5 /srv/crm-project/crm/sites/all/modules/civicrm/api/v3/System.php(50): CRM_Core_Invoke::rebuildMenuAndCaches(true, false)
#6 /srv/crm-project/crm/sites/all/modules/civicrm/Civi/API/Provider/MagicFunctionProvider.php(89): civicrm_api3_system_flush(Array)
#7 /srv/crm-project/crm/sites/all/modules/civicrm/Civi/API/Kernel.php(174): Civi\API\Provider\MagicFunctionProvider->invoke(Array)
#8 /srv/crm-project/crm/sites/all/modules/civicrm/Civi/API/Kernel.php(100): Civi\API\Kernel->runRequest(Array)
#9 /srv/crm-project/crm/sites/all/modules/civicrm/api/api.php(23): Civi\API\Kernel->runSafe('System', 'flush', Array, NULL)
#10 phar:///usr/local/bin/cv/src/Command/BaseCommand.php(102): civicrm_api('System', 'flush', Array)
#11 phar:///usr/local/bin/cv/src/Command/FlushCommand.php(37): Civi\Cv\Command\BaseCommand->callApiSuccess(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput), 'System', 'flush', Array)
#12 phar:///usr/local/bin/cv/vendor/symfony/console/Command/Command.php(257): Civi\Cv\Command\FlushCommand->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#13 phar:///usr/local/bin/cv/vendor/symfony/console/Application.php(850): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#14 phar:///usr/local/bin/cv/vendor/symfony/console/Application.php(193): Symfony\Component\Console\Application->doRunCommand(Object(Civi\Cv\Command\FlushCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#15 phar:///usr/local/bin/cv/vendor/symfony/console/Application.php(124): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#16 phar:///usr/local/bin/cv/src/Application.php(17): Symfony\Component\Console\Application->run()
#17 phar:///usr/local/bin/cv/bin/cv(22): Civi\Cv\Application::main('phar:///usr/loc...')
#18 /usr/local/bin/cv(10): require('phar:///usr/loc...')
#19 {main}
            [is_error] => 1
            [error_message] => API error: There is already a trigger for this class_name or combination of object_name and op
        )

)
```

with a little debugging on this line: ` /srv/crm-project/crm/sites/all/modules/civicrm/CRM/Core/ManagedEntities.php(234)`:

```php
  public function insertNewEntity($todo) {
    var_export($todo);
    $result = civicrm_api($todo['entity'], 'create', $todo['params']);
```

I can get this output:

```php
array (
  'name' => 'Civirules:Trigger.ContactCustomDataChanged',
  'entity' => 'CiviRuleTrigger',
  'params' => 
  array (
    'version' => 3,
    'name' => 'changed_contact_custom_data',
    'label' => 'Custom data on Contact (of any type) changed',
    'cron' => 0,
    'class_name' => 'CRM_CivirulesPostTrigger_ContactCustomDataChanged',
    'is_active' => 1,
  ),
  'module' => 'org.civicoop.civirules',
)
```

any insight into what this problem is?

update: this happens through the GUI as well. disabling civirules solves nothing. any tips on how to purge civirules so extensions start working again?

**Comment by lsh-0** [comment]
as per this ticket: 
https://github.com/CiviCooP/org.civicoop.civirules/issues/117

this works:
`delete from civicrm_managed where module = 'org.civicoop.civirules';`

and lets me re-enable extensions again.

**Comment by lsh-0** [comment]
(note: the above only works if the extension is disabled)

**Comment by jaapjansma** [comment]
I dont know what the problem is. There seems something with managed entities not working right.

**Comment by lsh-0** [comment]
my work is happening here: https://github.com/CiviCooP/org.civicoop.civirules/compare/master...elifesciences:master

I'm not sure if it will be useful to you guys and there may be more to come, but it appears to be working without issue

**Comment by ErikHommel** [comment]
We have fixed this in #189. The fix will be included from release 2.0 (expected 13 June 2018)
